doc/builders/builders.chapter.md

```diff
@@ -0,0 +1,42 @@
+# The builders {#chap-builders}
+
+A Builder, in the context of Nixpkgs, are functions that produces derivations. (Don't confuse it with the `builder` input argument of function `derivation`, which refers to the path to the executable that builds the derivation.) Such function is usually designed to follow a specific workflow, so that one can produce a package by specifying a few options instead of messing up with the `derivation` function.
+
+`stdenv.mkDerivation` is the most widely used builder, which is also the base of many other builders. It defaults to a workflow based on C and Makefile, but offers various options to customize parts of the builds. It's usage is See [](#sec-using-stdenv) for its usage in detail.
+
+There's no uniform standard about the input type of the builders. The trivial builders tends to have various input types for convenience, but language- or framework-specific builders tends to follow the style of `stdenv.mkDerivation`, which accepts an attribute set or a fixed-point function of attribute set.
+
+## Recursive attribute support in buirders {#sec-builders-recursive}
+
+The builder that accepts a fixed-point function is known to support recursive attributes. As [`stdenv.mkDerivation`'s support of recursive attributes](#mkderivation-recursive-attributes) was first included in the Nixpkgs release 22.05, `mkDerivation`-style buirders don't nessesarily support recursive attributes.
+
+The Nixpkgs Library function `lib.extendRecursiveBuilder` can be used to define a builders with recursive attributes support from an existing one. Suppose we would like to define a builder `mkLocalDerivaiton`, we could write the following in `pkgs/path/to/local-derivation/default.nix`:
+
+```nix
+{ lib
+, stdenv
+}:
+
+lib.extendRecursiveBuilder stdenv.mkDerivation [ ] (finalAttrs:
+
+{ preferLocalBuild ? true
+, allowSubstitute ? false
+}@args:
+
+args // {
+  inherit
+    preferLocalBuild
+    allowSubstitute
+    ;
+})
```

Comment:
````suggestion
```nix
{ lib, stdenv }:

lib.extendRecursiveBuilder stdenv.mkDerivation [ ] (finalAttrs:

  { preferLocalBuild ? true, allowSubstitute ? false }@args:
    args // { inherit preferLocalBuild allowSubstitute; }

)
````

For the sake of avoiding the example to look scary to inexperienced Nix language users (which are the majority of readers), I propose to collapse lines insignificant to the demonstration.

Reply:
The "in Nixpkgs" part means to demonstrates its ability to keep the existing indentation.

I'll simply remove it here to make it shorter, as it is already presented in the `Recursive attribute support in builders` section of the Nixpkgs manual.